### ISSUE TITLE
why did not my custom objective function work？

### ISSUE THREAD

**Issue by wangbingnan136** [microsoft/LightGBM#2052]
here is my custome objective function，I would like to add the ‘focal loss’ to lightgbm。

```
##################### my code################
from numpy import log
def grad_func(y_true,y_pred):
    if y_true==1:
        return -0.5*(2*y_pred - 2)*log(y_pred) - 0.5*(-y_pred + 1)**2/y_pred
    if y_true==0:
        return 0.5*y_pred**2/(-y_pred + 1) - 1.0*y_pred*log(-y_pred + 1)

def hess_func(y_true,y_pred):
    if y_true==1:
        return -1.0*log(y_pred) - (2.0*y_pred - 2.0)/y_pred + 0.5*(y_pred - 1)**2/y_pred**2
    if y_true==0:
        return 0.5*y_pred**2/(y_pred - 1)**2 - 2.0*y_pred/(y_pred - 1) - 1.0*log(-y_pred + 1)

def focal_loss(y_true,y_pred):
    gamma=2.0
    alpha=0.5
    y_true=y_true.tolist()
    y_pred=y_pred.tolist()
    grad=[]
    hess=[]
    for i in range(len(y_true)):
        if y_pred[i]==0:
            y_pred[i]=y_pred[i]+0.000000000001 ## avoid that the y_pred==0
#and can not be divided

        grad.append(grad_func(y_true[i],y_pred[i]))
        hess.append(hess_func(y_true[i],y_pred[i]))
    return np.array(grad),np.array(hess)

######################################################
```
It will run but always likes this
![image](https://user-images.githubusercontent.com/42114835/54510362-ee5d1100-4987-11e9-90fc-5028e1cba8d4.png)

**Comment by StrikerRUS** [comment]
Something should be wrong with calculations of grad/hess in your implementation. You can take a look at the following repo for XGBoost with focal loss: https://github.com/jhwjhw0123/Imbalance-XGBoost/blob/master/imxgboost/focal_loss.py.
